### PR TITLE
OCS-576: Prometheus pod restart should not have any functional impact

### PR DESCRIPTION
**Pull request by Akarsha-rai**
A testcase to validate prometheus pod restart should not have any functional impact, i.e the data/metrics shouldn't be lost after the restart of prometheus pod.

Signed-off-by: Akarsha <akrai@redhat.com>

### CODE REVIEW

**Comment by Akarsha-rai** [comment]
[test_log.txt](https://github.com/red-hat-storage/ocs-ci/files/3477046/test_log.txt)

============================================================================ 
  short test summary info ============================================================================
PASSED tests/e2e/monitoring/test_prometheus_pod_restart.py::TestRespinPrometheusPod::test_respinning_ceph_pods_and_interaction_with_prometheus_pod
============================================================================ 1 passed in 404.01 seconds ============================================================================

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    # Delete projects created
+    for prj in self.namespace_list:
+        prj_obj = ocp.OCP(kind='Project', namespace=prj)
+        prj_obj.delete(resource_name=prj)
+
+
+def create_project_and_pvc_and_check_metrics_are_collected(self):
+    """
+    Creates projects, pvcs and app pods
+
```
Comment (RazTamir) [inline tests/e2e/monitoring/test_prometheus_pod_restart.py]: missing return value.
There are asserts and also return True at the end - you need to ensure the function is returning True/False or raising assertion. doing both can create confution

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        # Get the pvc which mounted on prometheus[0] pod
+        pod_info = pod_obj[0].get()
+        pvc_name = pod_info['spec']['volumes'][0]['persistentVolumeClaim']['claimName']
+
+        # Check the used space of prometheus[0] pod
+        initial_used_size = check_used_size_of_prometheus_pod(pod_obj[0])
+
+        # Respin one of the prometheus pod
+        pod_obj[0].delete(force=True)
+        POD = ocp.OCP(kind=constants.POD, namespace=defaults.OCS_MONITORING_NAMESPACE)
```
Comment (RazTamir) [inline tests/e2e/monitoring/test_prometheus_pod_restart.py]: POD should be pod

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        assert POD.wait_for_resource(
+            condition='Running', selector=f'app=prometheus', timeout=300
+        )
+
+        # Check the same pvc is mounted on new pod
+        pod_info = pod_obj[0].get()
+        assert pod_info['spec']['volumes'][0]['persistentVolumeClaim']['claimName'] in pvc_name
```
Comment (RazTamir) [inline tests/e2e/monitoring/test_prometheus_pod_restart.py]: For each assert statement, add assertion log:
```python
assert False, "Failed to get pod...."
```

File: ocs_ci/ocs/resources/pod.py [hunk 4]
```diff
@@ -592,8 +601,25 @@ def get_pod_obj(name, namespace=None):
...
         obj : A pod object
+
     """
     ocp_obj = OCP(api_version='v1', kind=constants.POD, namespace=namespace)
     ocp_dict = ocp_obj.get(resource_name=name)
     pod_obj = Pod(**ocp_dict)
     return pod_obj
+
+
+def delete_pods(pod_objs):
```
Comment (keesturam) [inline ocs_ci/ocs/resources/pod.py]: This function assumes "namespace". pls, fetch namespace as an argument for this function.

File: ocs_ci/ocs/monitoring.py [hunk 5]
```diff
@@ -53,49 +59,83 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
     )
     pvc_obj_list = []
     for pvc in pvc_list['items']:
         pvc_dict = ocp_pvc_obj.get(resource_name=pvc.get('metadata').get('name'))
         pvc_obj = PVC(**pvc_dict)
         pvc_obj_list.append(pvc_obj)
     return pvc_obj_list
+
+
+def get_kube_pod_spec_volumes_persistentvolumeclaims_info_metric():
```
Comment (keesturam) [inline ocs_ci/ocs/monitoring.py]: function name looks too long and I found it difficult to understand. Consider changing the name to a smaller name. for eg., get_metrics_persistentvolumeclaims_info.

**Comment by vasukulkarni** [comment]
@Akarsha-rai Nice test case, please resolve conflicts and lets try to get this merged soon.

File: tests/e2e/monitoring/test_prometheus_pod_restart.py (hunk 1 above)
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: removed this function and creating separately

File: tests/e2e/monitoring/test_prometheus_pod_restart.py (hunk 2 above)
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: done

File: tests/e2e/monitoring/test_prometheus_pod_restart.py (hunk 3 above)
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: done

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+)
+@pytest.mark.polarion_id("OCS-576")
+class TestRespinPrometheusPod(E2ETest):
+    """
+    Prometheus pod restart should not have any functional impact,
+    i.e the data/metrics shouldn't be lost after the restart of prometheus pod.
+    """
+
+    @tier4
+    def test_respinning_ceph_pods_and_interaction_with_prometheus_pod(self):
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: changed:)

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        pvc_name = pod_info['spec']['volumes'][0]['persistentVolumeClaim']['claimName']
+
+        # Check the used space of prometheus[0] pod
+        initial_used_size = check_used_size_of_prometheus_pod(pod_obj[0])
+
+        # Respin one of the prometheus pod
+        pod_obj[0].delete(force=True)
+        POD = ocp.OCP(kind=constants.POD, namespace=defaults.OCS_MONITORING_NAMESPACE)
+        assert POD.wait_for_resource(
+            condition='Running', selector=f'app=prometheus', timeout=300
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: change to 60

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        )
+
+        # Check the same pvc is mounted on new pod
+        pod_info = pod_obj[0].get()
+        assert pod_info['spec']['volumes'][0]['persistentVolumeClaim']['claimName'] in pvc_name
+
+        # Check the used space of prometheus[0] pod should be equal or increased after respinning pod
+        # i.e, the data/metrics should not be deleted which was collected before
+        used_size_after_respin = check_used_size_of_prometheus_pod(pod_obj[0])
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: removed:)

File: tests/e2e/monitoring/test_prometheus_pod_restart.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        Test case to validate prometheus pod restart
+        should not have any functional impact
+        """
+
+        # Get the prometheus pod
+        pod_obj = pod.get_all_pods(namespace=defaults.OCS_MONITORING_NAMESPACE, selector=['prometheus'])
+
+        # Get the pvc which mounted on prometheus[0] pod
+        pod_info = pod_obj[0].get()
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_prometheus_pod_restart.py]: done

**Comment by Akarsha-rai** [comment]
[OCS-576.txt](https://github.com/red-hat-storage/ocs-ci/files/3559584/OCS-576.txt)


PASSED tests/e2e/monitoring/test_prometheus_pod_restart.py::TestPrometheusPodRestart::test_monitoring_after_restarting_prometheus_pod
=============================================================================1 passed in 258.31 seconds =============================================================================

**Comment by vasukulkarni** [comment]
Needs rebase

**Comment by Akarsha-rai** [comment]
> Needs rebase

done

**Comment by Akarsha-rai** [comment]
[OCS-576.txt](https://github.com/red-hat-storage/ocs-ci/files/3739067/OCS-576.txt)

=============================================================================short test summary info =============================================================================
PASSED tests/e2e/monitoring/test_prometheus_pod_restart.py::TestPrometheusPodRestart::test_monitoring_after_restarting_prometheus_pod
=========================================================================================== 1 passed in 338.28 seconds =============================================================================

**Comment by Akarsha-rai** [comment]
Since this to old PR and changes need to be done.Closing this PR.
Sent a new PR to address this test case #1228 .

**Pull request closed without merging**